Fix(html): Handle `<br>` elements to insert line breaks in text

Dhruv-Maradiya opened the pull request:
Fixes #1090 by updating the DOM parser to handle `<br>` elements and insert line breaks (`\n`) when converting HTML content to plain text.

Initially, I thought adding a simple condition might not be a reliable solution. So, I decided to check how HTML-to-text conversion is handled in Chromium and found a similar approach. Here's the [link](https://chromium.googlesource.com/chromium/blink/+/refs/heads/main/Source/core/dom/Node.cpp#1244). 


---

- [x] I’ve reviewed the contributor guide and applied the relevant portions to this PR.

<details>
  <summary>Contribution guidelines:</summary><br>

- See our [contributor guide](https://github.com/dart-lang/.github/blob/main/CONTRIBUTING.md) for general expectations for PRs.
- Larger or significant changes should be discussed in an issue before creating a PR.
- Contributions to our repos should follow the [Dart style guide](https://dart.dev/guides/language/effective-dart) and use `dart format`.
- Most changes should add an entry to the changelog and may need to [rev the pubspec package version](https://github.com/dart-lang/sdk/blob/main/docs/External-Package-Maintenance.md#making-a-change).
- Changes to packages require [corresponding tests](https://github.com/dart-lang/.github/blob/main/CONTRIBUTING.md#Testing).

Note that many Dart repos have a weekly cadence for reviewing PRs - please allow for some latency before initial review feedback.
</details>

Dhruv-Maradiya committed:
Implement handling of <br> elements in DOM parsing to insert line breaks in text

Dhruv-Maradiya committed:
update tests to handle <br> elements for line breaks

Dhruv-Maradiya committed:
update changelog to include fix for <br> newline handling

Dhruv-Maradiya committed:
bump version to 0.15.5+1 in html package

HosseinYousefi commented on pkgs/html/lib/dom_parsing.dart:
```diff
@@ -43,7 +43,14 @@ class TreeVisitor {
   // TODO(jmesserly): visit attributes.
-  void visitElement(Element node) => visitNodeFallback(node);
+  void visitElement(Element node) {
+    if (node.localName == 'br') {
```
This can be separate `isElementBr` function. Is this if statement enough to recognize it?

HosseinYousefi commented on pkgs/html/test/parser_feature_test.dart:
```diff
@@ -265,11 +265,11 @@ On line 4, column 3 of ParseError: Unexpected DOCTYPE. Ignored.
     final text = e.firstChild!;
     expect((text as Text).data, 'foo');
-    expect(e.text, 'foobarbaz');
+    expect(e.text, 'foo\nbarbaz\n');
```
I wouldn't change the behavior of `.text`, in the link you shared we had a function with an explicit boolean like `textContent({bool convertBRsToNewlines = false})`, I'd prefer to add that instead.

Dhruv-Maradiya commented:
Hey, thanks for reviewing this! 🙌
It’s been a few months since I worked on it, and I was still getting familiar with the codebase at the time — so I’ll need to refresh myself on the changes.
I'll take a look as soon as I can. Appreciate your feedback!

Dhruv-Maradiya committed:
feat: add spec-compliant textContent with BR conversion flag

Implements DOM spec textContent algorithm with optional convertBRsToNewlines parameter. Adds isElementBr() helper for namespace-aware BR detection. Maintains backward compatibility with existing .text getter.

Dhruv-Maradiya committed:
Merge branch 'main' into main

HosseinYousefi commented on pkgs/html/CHANGELOG.md:
```diff
@@ -1,6 +1,7 @@
 ## 0.15.7-wip
 
 - Require Dart `3.6`
+- Fix a bug in DOM parsing where `<br>` tags does not create a new line when html is converted to text.
```
It's no longer a bug then, change the changelog entry to convey adding a new `textContent` method instead.

Dhruv-Maradiya committed:
feat: add textContent method and fix <br> tag handling in DOM parsing
